Feat: Cuda Device Properties

micahcc opened the pull request:
- Add function to get device properties
- There was one breaking change in the cudaDeviceProps struct between cuda 10 and 11 (they inserted persistingL2CacheMaxSize), but other than that they have only added. 
- Its nice to have the type defined in rust, because its easier for users, but it may be worth generating the type with bindgen if it gets changed again.

Different versions of cudaDeviceProps struct:
<img width="1914" height="1542" alt="2025-08-26T11-29-49" src="https://github.com/user-attachments/assets/a1ebead8-b546-4c82-afb4-b6a29ff96566" />

Commit:
add pci stuff

Commit:
this is better

Commit:
make name u8

Commit:
Revert "make name u8"

This reverts commit 9478726aa7268c58bea80a80cdc88a624a19f418.

Commit:
Reapply "make name u8"

This reverts commit e245e6f9175707078e1264490bc72df6c02523df.

Commit:
add missing

Commit:
add missing

gerwin3 commented:
Oh so the version in your PR, to which version does it correspond in CUDA exactly?

EDIT: @micahcc And could you also add the padding to it? I think technically without it the struct may be incompatible due to alignment. Also a static check on the size of the struct would be nice :)

micahcc commented:
@gerwin3 I think without codegen this might get pretty ugly. I think we would want to make a C version of the cudaDeviceProperties with our own layout and then manually copy each field. That way we can guarantee the memory layout in rust. The cuda versions are listed in the diff view I posted (10, 11.8, 12.8).

I'm fine doing the grunt work of doing the manual copy if you're ok with the long term maintenance burden of it (I don't think its high, but some people might balk)

Honestly though the memcpy I have right now will handle any lengthening or shortening of the struct, the only risk is if they add something in the middle, but nothing should result in buffer overruns, at worst it would lead to corrupt data.

Commit:
add device tensor

Commit:
add debug for ptr

Commit:
as ptr

Commit:
type

gerwin3 commented:
Actually as shown in your diff the CUDA team has already taken this into account. They added padding to the end of the struct to ensure it is the same size across versions (it is basically a promise they will never go bigger than that). You can just embed the padding on the Rust-side. As long as the struct has `repr(C)` that will ensure the types are compatible regardless of any changes on the CUDA side.

gerwin3 commented:
Also, could you split this PR in 2:

- Device props
- DeviceTensor

That way I can review and accept them separately.

micahcc commented:
Sorry yeah, shouldn't have merged main, let me make branches

Commit:
remove, will make a separate PR

Commit:
cleanup

Commit:
add padding

Commit:
fix test

Commit:
sanity check

micahcc commented:
@gerwin3 ok, updated.
- Added more testing
- Added up to the padding from 12.6
- Removed the tensor changes (can make a new PR, that is for some TensorRT stuff)
